### PR TITLE
Pool lexed token arrays in SyntaxParser.

### PR DESCRIPTION
**Pull request by ToddGrun**
ArrayElement<SyntaxToken>[] accounts for 2.3% of allocations in the customer's trace I'm looking at. These changes should prevent the ctor and PreLex calls from allocating (1.8% combined). Handled very similarly to how the class already handles it's BlendedNode[].

image

### CODE REVIEW

**Review by sharwell — CHANGES_REQUESTED** [review]
Requesting an extended microbenchmark before merging this. This change appears to cause the core lexer to add a write barrier to each token due to relocation of all storage from Gen 0 to Gen 2.

**Comment by ToddGrun** [comment]
> Requesting an extended microbenchmark before merging this. This change appears to cause the core lexer to add a write barrier to each token due to relocation of all storage from Gen 0 to Gen 2.

Can you point me to some documentation outlining the write barrier that is added when this happens? I'm not quite clear on how that would differ from any of the pooling that we do. Is there any existing benchmarks in place that I can start with for testing this?

**Comment by CyrusNajmabadi** [comment]
@ToddGrun  for parsing, i would jsut write a benchmark.net case that just lexes and/or parses some file the fits within the 4k token limit, and one which has more htan 4k tokens.  woudl likely reveal immediately if there are issues.

**Comment by ToddGrun** [comment]
Benchmarking results (benchmarking code added in 3rd commit, removed in 4th):

```
|     Method | UsePooling |       Mean |    Error |   StdDev |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|----------- |----------- |-----------:|---------:|---------:|-------:|-------:|------:|----------:|
|  SmallFile |      False |   203.9 ns |  4.03 ns |  5.24 ns | 0.2456 | 0.0019 |     - |   2,568 B |
| MediumFile |      False | 1,039.8 ns | 20.58 ns | 25.28 ns | 3.1948 | 0.3185 |     - |  33,560 B |
|  LargeFile |      False | 1,051.8 ns | 11.24 ns |  9.38 ns | 3.1948 | 0.3185 |     - |  33,560 B |
|  SmallFile |       True |   116.4 ns |  1.76 ns |  1.64 ns | 0.0465 |      - |     - |     488 B |
| MediumFile |       True |   323.1 ns |  1.59 ns |  1.41 ns | 0.0463 |      - |     - |     488 B |
|  LargeFile |       True |   114.2 ns |  0.47 ns |  0.37 ns | 0.0466 |      - |     - |     488 B |
```

**Comment by ToddGrun** [comment]
@jaredpar @333fred ptal

**Comment by jaredpar** [comment]
@333fred has been doing the most work in this area. I'm a bit reluctant to approve in his abscence cause I don't want to inadvertently subvert anythhing he's working on. Will have him look when he's back.

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:146]: >Allocate() [](http://example.com/codeflow?start=47&length=10)

Can we pass the `size` hint?

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:104]: I don't have a clear sense of how to choose this magic number. I guess this is fine and we can tweak later if needed.

**Review by jcouv — DISMISSED** [review]
LGTM Thanks (iteration 4). We'll wait for Fred for second review.

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:104]: Some back-of-the-napkin math suggests that, at a guesstimated 6-10 tokens/line, this will cover C# files around 4-600 lines in length. That does feel somewhat low to me for real-world scenarios (since I think I might be low-balling the number of tokens/line on average), but I agree we can tweak if needed. Might be worth a comment indicating such though.

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:443]: I'm not sure why we would not attempt to retrieve from the pool in this scenario, or to discard the old array. Surely the intermediately-sized arrays have value, in that we need to work our way up through them to get to larger arrays? In other words, for one particularly large file, I'd expect us to populate the pool with all the intermediately-sized arrays (up to the max length we track), benefiting all other non-large files of those intermediate sizes.

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 4)

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:443]: I don't believe that the ObjectPool class being used has an Allocate method that takes in a size hint. So, if we did use the Allocate method to obtain a new array, there is no guarantee that it would be large enough for our needs. I suppose we could not forget the old array, but the value of doing that seems questionable. As an aside, the approach taken here closely mimics what the AddTokenSlot method does directly above.

**Comment by ToddGrun** [comment]
@333fred -- any remaining concerns?

**Comment by 333fred** [comment]
@ToddGrun I'm still not a fan of throwing away the intermediate arrays. They are likely quite useful, and it means that a large file can throw away a number of arrays . If `ObjectPool` has no way for us to rent an array of at least a specific size, we may want to consider just using `ArrayPool` from the BCL, which does have that ability.

**Comment by ToddGrun** [comment]
@333fred -- I don't see any uses of the BCL ArrayPool class throughout Roslyn (and it appears to only be available in >= 2.1), so I'm a bit wary of attempting that path. 

There is a Roslyn.Utiliities ArrayPool class that has a Rent method that takes in a minimum size, but I'm not clear on the performance characteristics of using that class vs an ObjectPool of arrays as is currently coded. @sharwell -- do you have a general preference between these two types, as I tend to see more of the former than the latter throughout the roslyn codebase.

**Comment by CyrusNajmabadi** [comment]
Two thoughts:

1. i think it's fine to toss the intermediary arrays.  there should only be a small number of those because of the doubling-scalign we're doing.
2. instead of this, we could just *start* the arrays at 4096 elements to begin with, avoiding small arrays iin the first place.  Those pooled arrays will still be tiny (4k pointers is nothing), and will then work fine for the small file case or hte large file case.  Then we never have to deal with intermediary arrays of small length.

**Comment by 333fred** [comment]
I would be fine with that approach Cyrus.

**Comment by ToddGrun** [comment]
Seems reasonable to me. I've started the array size at MaxCachedTokenArraySize

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:108]: Should this not be `==`?

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:448]: We want to discard if too large and otherwise clear/return the array now, right?

Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs]: Yeah, made a pass through the code again and did some cleanup including this

**Comment by ToddGrun** [comment]
Re-ran benchmarks against commit 7 (on my home machine this time) and still looking good:

```|     Method | UsePooling |       Mean |    Error |   StdDev |     Median |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|----------- |----------- |-----------:|---------:|---------:|-----------:|-------:|-------:|------:|----------:|
|  SmallFile |      False |   246.1 ns |  2.05 ns |  1.60 ns |   245.7 ns | 0.5956 |      - |     - |   2,568 B |
| MediumFile |      False | 1,704.8 ns | 22.75 ns | 31.14 ns | 1,697.9 ns | 7.7515 | 0.7744 |     - |  33,560 B |
|  LargeFile |      False | 1,693.7 ns |  8.43 ns | 18.51 ns | 1,686.6 ns | 7.7515 | 0.7744 |     - |  33,560 B |
|  SmallFile |       True |   147.3 ns |  1.85 ns |  1.64 ns |   146.9 ns | 0.1130 |      - |     - |     488 B |
| MediumFile |       True |   146.3 ns |  0.42 ns |  0.35 ns |   146.4 ns | 0.1130 |      - |     - |     488 B |
|  LargeFile |       True |   150.1 ns |  0.58 ns |  0.52 ns |   149.9 ns | 0.1130 |      - |     - |     488 B |

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:447]: What am I missing in this code? Would this not clear `_lexedTokens`, `Free` it, and then resize the empty array, throwing away all the tokens that have been lexed so far?

Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:447]: I did debug my changes, but I wasn't able to hit the else clauses here (and the if condition was complicated enough that I didn't want to mess with it)

**Comment by ToddGrun** [comment]
>                     Array.Copy(_lexedTokens, shiftOffset, _lexedTokens, 0, shiftCount);

I think I introduced a bug by adding the assumption to only clear what's after _tokenOffset in Dispose. Going to debug and likely push out an update soon.

---
Refers to: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:433 in ecf6906. [](commit_id = ecf6906c8303e1a1d4dd1539c83b6f8d2fa43fd0, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:436]: I don't think I understand why we need to do this. I don't think it's _wrong_, per se, but I'm not following what in the new logic would require this clear now when it wasn't before.

Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:436]: It needs to be done now because the Dispose method only does the Array.Clear from 0 -> _tokenCount, not the whole array before releasing the array back into the pool.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:436]: > because the Dispose method only does the Array.Clear from 0 -> _tokenCount

What's the reasoning behind not clearing the whole array? In theory, `Array.Clear` should just turn into `memset`, which I wouldn't expect to take very long, even on a 4KB array.

Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:436]: The benchmark for the small file case showed a significant regression  if I did the Array.Clear over the whole array.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:436]: Gotcha, thanks for the clarification.

**Review by 333fred — APPROVED** [review]

**Comment by 333fred** [comment]
@jcouv could you please take another look? There have been several changes since you signed off.

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:436]: The manipulations done around _tokenCount and _tokenOffset are complicated enough that I don't feel completely comfortable with assuming that _tokenCount is the proper length to clear out. (I hadn't seen it's manipulation in Reinitialize and how it can be used to track into _blendedTokens too). Instead, I think it will be clearer and safer to just keep track of the largest index written into the lexedTokenArray, and just clear that amount out on dispose.

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:118]: Should we `ForgetTrackedObject` in the alternative (not putting lexedTokens back into the pool)?

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:461]: In `Dispose` we clear up to `_maxWrittenLexedTokenIndex`, but here we clear the entire length. Why the difference?

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:463]: Same question here (should we forget the tracking in the alternative).

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 11)

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:118]: If it's not equal to this size, then it didn't come from the pool (as the pool will now only ever contain arrays of this length).

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:463]: see other comment

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:461]: There's no real reason other than I believe if we hit this codepath, then the two values are probably the same. But, it's clearer if we're consistent between the two places doing this, so I'll change this to match.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs]: Thanks. Consider factoring/extracting this common code

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 12) with a nit/factoring to consider. We'll wait for your response (up to you) before merging/squashing.

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs]: Very reasonable request now that they are the same.

**Comment by ToddGrun** [comment]
Made the final recommended change. Thanks for all the feedback, let me know if you want me to go ahead and merge or if you want to push the button.

**Comment by 333fred** [comment]
Go for it

**Comment by jcouv** [comment]
I went ahead and merged/squashed since everything was green. Thanks for the contribution @ToddGrun !

**Pull request merged**
